### DISCUSSION TITLE
Why does the docsite state that position: absolute is the more performant strategy?

### DISCUSSION THREAD

**Discussion by ling1726** [floating-ui/floating-ui#2354] [Q&A]
Hi,

Would like to ask why the Floating UI docsite states that position: absolute is the more performant strategy for the browser? This statement has determined a lot of our design system's choices when it comes to using Floating UI.

However, recently we have noticed that position: fixed would be a far 'simpler' solution since it creates its own stacking context and handles overflow a lot better when dealing with nested floating elements.

I'd like to know if there's been any past experience or testing that motivated this statement. I tried to run a few searches online and failed to find other mentions of performance between absolute and fixed position. If there is some material that I'm missing, it might be nice to reference it in the docsite to avoid future discussions like this or https://github.com/floating-ui/floating-ui/discussions/1777. The current formulation seems like a very sweeping statement, especially without other supporting material

![image](https://github.com/floating-ui/floating-ui/assets/20744592/fc88eb72-c271-405c-8068-db340efe5771)

**Comment by atomiks** [comment]
In general, `absolute` has fewer quirks with anchoring (the overscroll problem mentioned in the issue), and usually needs to update and re-render less frequently (scrolling context differences). 

Further, `fixed` seems to have so many bugs that keep popping up that need to be handled and adjusted for. Trying to detect a containing block, for instance, has broken when browsers change their internal implementation without notice (and there's no way to feature detect it properly). In Safari we have to use a hack to get it to even position itself correctly when pinch-zooming. GitHub doesn't even do this and is broken (e.g. the timeline on user profiles).

Overall, I don't see any reason to use `fixed` **unless** the reference element is in a fixed container (or fixed itself), but the floating element is not (e.g. portaled) — or you absolutely need to keep DOM context in-tact and can't portal the floating element. But since you mentioned stacking contexts, maybe there are some things I'm not aware of.

**Reply by ling1726** [reply]
I think the main reason for my question with `fixed` is that in Fluent UI we support a lot of nested floating use cases. If we could avoid portalling to break the stacking context it might give the following advantages:

1. Accessibility - portalling floating elements isn't a great experience for screen reader users because in virtual cursor mode they will end up at the end of the document which is confusing if the button that opened the floating element is somewhere else
2. Nested floating elements - If you create menus that open other menus, you _**must**_ portal each menu. In nested DOM, if a parent menu overflows and displays a scrollbar, it will overflow all its child menus. This doesn't happen when using position: fixed and using top/left instead of transforms.
![image](https://github.com/floating-ui/floating-ui/assets/20744592/47b311ad-e60d-462d-a45a-9f24a41d98bc)

3. Maintaining event propagation in nested floating elements - when you portal you lose event propagation and can't use `element.contains` to test that one floating element is in another. in Fluent UI we've resorted to using [virtual parents](https://github.com/microsoft/fluentui/tree/master/packages/react-components/react-portal#virtual-parents). It just makes creating things like nested menus all that harder

I don't doubt there's good reasons you know why `fixed` is not a great choice, and your reply you've given quite a few. Perhaps it might be nice to incorporate them into the docs so that users can see some of the reasoning behind the choices

**Reply by atomiks** [reply]
1. I agree, but it's possible to make work - I made it work with [`FloatingFocusManager`](https://floating-ui.com/docs/FloatingFocusManager) which handles virtual cursor focus management with `aria-owns` (most screen readers) + focus guards (macOS + iOS).

2. Using `fixed` for submenus only is a decent trade-off, but if you try to position the floating element using `transform`, or add `backdrop-filter` to the floating element, or a `transform` animation - it creates a containing block and then clips it. Portaling is the only 100% reliable method (aside for the new top layer API)

3. Fluent UI uses React right? React's event delegation system bubbles through portals, so you can detect if a click occurred inside the React tree, unlike the DOM tree. And [`FloatingTree`](https://floating-ui.com/docs/FloatingTree) handles other cases in this situation too.

**Reply by ling1726** [reply]
> aria-owns

Depending on the screen reader we've gotten some very weird results for this. Granted it's been a year or two since we've tested it so it might be worth looking into again

> Portaling is the only 100% reliable method (aside for the new top layer API)
>  so you can detect if a click occurred inside the React tree

Yeah agreed that portalling is 100% method but it generates some unnecessary complexity. The React tree will handle most event bubbles, but if you want to dismiss a floating element on a click smwhere on document the event listener should be on the document. At that point we hit the `element.contains` problem (everything is portalled) and had to figure out that virtual parent way of getting around it, would the FloatingTree handle that?

**Reply by atomiks** [reply]
In a fully generic design system, unfortunately I can only recommend always portaling, because if they try to add `backdrop-filter` or something else, suddenly it breaks. It needs to robustly handle every possible case as I'm sure you know of 😢 

Can you describe the `contains` problem more? One thing the React tree event system can't handle that `FloatingTree` does, is nested outside press dismissals if you want them not to all close simultaneously.

**Reply by ling1726** [reply]
> In a fully generic design system, unfortunately I can only recommend always portaling, because if they try to add backdrop-filter or something else, suddenly it breaks. It needs to robustly handle every possible case as I'm sure you know of 😢

Yeah, good to know that the pain is shared 😭😭 

> nested outside press dismissals if you want them not to all close simultaneously.

I think we might be talking about the same problem. I'll describe mine a bit better anyway and let's see. Essentially, when you portal nested floating elements (i.e. menus) and you want to dismiss all the floating elements when someone clicks outside the stack, you add an event listener to `document`.

However since all the floating elements are all portals, you can't use `element.contains` to check if the click came from a child. In the below example, if I click on the outmost popover, I don't want to dismiss the parents.

![image](https://github.com/floating-ui/floating-ui/assets/20744592/f8b156e6-0512-4686-9179-96979290b6e3)

It's why we came up with the idea of [virtual DOM parents](https://github.com/microsoft/fluentui/tree/master/packages/react-components/react-portal#virtual-parents) and a custom `elementContains` imeplementation

**Reply by atomiks** [reply]
I think you're describing something that can be fixed with the React tree system. You can detect if a click occurred inside the React tree, before the document listener fires, using capture phase: https://github.com/floating-ui/floating-ui/pull/1852/files

However, apparently outside presses on the document should use capture phase too, but then it breaks that trick. Haven't found a solution to that yet: https://github.com/floating-ui/floating-ui/pull/2345

**Reply by ling1726** [reply]
But clicks that happen inside the React tree are valid too, if they are not a part of the stack of floating elements. How would you handle clicks inside the react tree then?

> However, apparently outside presses on the document should use capture phase too, but then it breaks that trick.

Sorry I didn't quite get that one, do you mind elaborating a bit?

**Reply by atomiks** [reply]
As long as it's not a direct ancestor, it won't be considered "inside". I haven't encountered a problem with it so far since that's not likely.

> Sorry I didn't quite get that one, do you mind elaborating a bit?

It relies on the React listener firing before the native `document` one. Since they both happen in capture phase, the `document` one fires first and the ref is no longer written in time before it runs (and early-returns).

**Reply by ling1726** [reply]
Ah gotcha, I think I get your solution with capture events in the react tree now, It's a pretty nifty solution for a React only app, and I'll probably have to remember it for other projects.

> It relies on the React listener firing before the native document one

Yes, I get it, we _**always**_ have to support multi React root and multi framework applications, so I think we never hit this problem, because we never considered that React could be a full solution to outside dismiss, so we went a bit lower level to the DOM on our end.

**Reply by atomiks** [reply]
The PR I linked links to the react-aria issue about it. If you're using their `button` component, it stops propagation or something, so it needs to use the earlier phase.

**Reply by FezVrasta** [reply]
Would the new popover API help or is that analogous to position fixed?
